File: wasi-js-bindings/test/clock.js
// Copyright 2020 The Chromium OS Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * @fileoverview Tests for clock APIs.
 */

import * as Process from '../js/process.js';
import * as SyscallEntry from '../js/syscall_entry.js';
import * as SyscallHandler from '../js/syscall_handler.js';
import * as WASI from '../js/wasi.js';

describe('clock.js', () => {

/**
 * A handler just to capture output.
 */
class TestSyscallHandler extends SyscallHandler.DirectWasiPreview1 {
  constructor(...args) {
    super(...args);
    this.stdout = '';
    this.stderr = '';
    this.td = new TextDecoder();
  }

  /** @override */
  handle_fd_write(fd, buf) {
    switch (fd) {
      case 1:
        this.stdout += this.td.decode(buf, {stream: true});
        return WASI.errno.ESUCCESS;

      case 2:
        this.stderr += this.td.decode(buf, {stream: true});
        return WASI.errno.ESUCCESS;
    }

    return WASI.errno.EINVAL;
  }
}

/**
 * The format the WASM program outputs.
 *
 * @typedef {{
 *   getres: !Array<string>,
 *   gettime: !Array<!Array<string>>,
 * }}
 */
const TestData = {};

/**
 * Helper function to run the wasm module & return output.
 *
 * @param {!ArrayBuffer} prog The program to run.
 * @param {!Array<string>} argv The program arguments.
 * @return {!Object} The program results.
 */
async function run(prog, argv) {
  const handler = new TestSyscallHandler();
  const sys_handlers = [handler];
  const proc = new Process.Foreground({
    executable: prog,
    argv: ['clock.wasm', ...argv],
    sys_handlers: sys_handlers,
    sys_entries: [
      new SyscallEntry.WasiPreview1({sys_handlers}),
    ],
  });
  const ret = await proc.run();
  assert.equal(handler.stderr, '');
  return {
    returncode: ret,
    stdout: handler.stdout,
    stderr: handler.stderr,
    data: /** @type {!TestData} */ (JSON.parse(handler.stdout)),
  };
}

/**
 * Load some common state that all tests in here want.
 */
before(async function() {
  /**
   * Fetch & read the body once to speed up the tests.
   *
   * @type {!ArrayBuffer}
   */
  this.prog = await fetch('clock.wasm')
    .then((response) => response.arrayBuffer());
});

/**
 * Combine timespec fields into a single value.
 *
 * @param {string} seconds The tv_sec[onds] field.
 * @param {string} nanoseconds The tv_n[ano]sec[onds] field.
 * @return {bigint} The time in nanoseconds.
 */
function getNanosec(seconds, nanoseconds) {
  return BigInt(seconds) * 1000000000n + BigInt(nanoseconds);
}

/**
 * Verify monotonic clock behavior.
 */
it('monotonic', async function() {
  const result = await run(this.prog, ['monotonic']);
  const data = result.data;

  // Check resolution.
  assert.deepEqual(data.getres, ['0', '1']);

  // Check it is indeed monotonic.
  let curr = -1n;
  for (let i = 0; i < data.gettime.length; ++i) {
    const next = getNanosec(...data.gettime[i]);
    // assert.isBelow doesn't support BigInt yet.
    assert(curr < next, `${curr} < ${next}`);
    curr = next;
  }
});

/**
 * Verify realtime clock behavior.
 */
it('realtime', async function() {
  const result = await run(this.prog, ['realtime']);
  const data = result.data;

  // Check resolution.
  assert.deepEqual(data.getres, ['0', '1000000']);

  // Check it is indeed monotonic.
  let curr = -1n;
  for (let i = 0; i < data.gettime.length; ++i) {
    const next = getNanosec(...data.gettime[i]);
    // assert.isBelow doesn't support BigInt yet.
    assert(curr < next, `${curr} < ${next}`);
    curr = next;
  }
});

});

File: wasi-js-bindings/test/clock.c
// Copyright 2020 The Chromium OS Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// Utility to dump clock info.

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "test-utils.h"

static void dump_clock(const char* name, clockid_t clk_id) {
  struct timespec ts;
  int ret;

  ret = clock_getres(clk_id, &ts);
  assert(ret == 0);
  // Print the numbers as strings to avoid JSON->JS issues with large values.
  printf("  \"getres\": [\"%lli\", \"%lli\"],\n",
         (long long)ts.tv_sec, (long long)ts.tv_nsec);

  // Sample it a few times to observe its behavior.
  printf("  \"gettime\": [\n");
  const int kIters = 5;
  for (int i = 0; i < kIters; ++i) {
    if (i)
      usleep(1000);

    ret = clock_gettime(clk_id, &ts);
    assert(ret == 0);
    printf("    [\"%lli\", \"%lli\"]%s\n",
           (long long)ts.tv_sec, (long long)ts.tv_nsec,
           i == kIters - 1 ? "" : ",");
  }
  printf("  ]\n");
}

int main(int argc, char *argv[]) {
  if (argc != 2) {
    fprintf(stderr, "Usage: clock <source>\n");
    abort();
  }

  const char* mode = argv[1];

  // Output in JSON format for easier test runner parsing.
  printf("{\n");

  if (streq(mode, "realtime")) {
    dump_clock("realtime", CLOCK_REALTIME);
  } else if (streq(mode, "monotonic")) {
    dump_clock("monotonic", CLOCK_MONOTONIC);
  } else {
    fprintf(stderr, "unknown mode '%s'\n", mode);
    abort();
  }

  printf("}\n");
  return 0;
}
